// typedoc/dist/lib/converter/types/anonymous.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ts = require("typescript");
const components_1 = require("../components");
const models_1 = require("../../models");
class ObjectType extends models_1.Type {
    constructor() {
        super(...arguments);
        this.type = 'object';
    }
    clone() {
        return new ObjectType;
    }
}
exports.ObjectType = ObjectType;
let ConditionalConverter = class ConditionalConverter extends components_1.ConverterTypeComponent {
    supportsType(context, type) {
        return !!(type.flags & ts.TypeFlags.Object)
            && !!(type.objectFlags & ts.ObjectFlags.Anonymous)
            && !type.symbol;
    }
    convertType(context, type) {
        return new ObjectType;
    }
};
ConditionalConverter = __decorate([
    components_1.Component({ name: 'type:anonymous' })
], ConditionalConverter);
exports.ConditionalConverter = ConditionalConverter;
